src/VisualStudio/Core/Def/Implementation/Experimentation/AnalyzerVsixSuggestedActionCallback.cs

```diff
@@ -0,0 +1,234 @@
+        private readonly IForegroundNotificationService _foregroundNotificationService;
+        private readonly IAsynchronousOperationListener _operationListener;
+        private readonly SVsServiceProvider _serviceProvider;
+
+        private readonly object _scheduleLock = new object();
+        // If we are currently checking to see if the user fits the treatment profile, then we don't want to schedule another check
+        private bool _checksRunning = false;
+        // We don't show the user an infobar more than once per VS session
+        private bool _infoBarShown = false;
+        private FxCopInstallStatus _installStatus = FxCopInstallStatus.Unchecked;
+        // Initialized the first time we're on the UI thread
+        private IExperimentationService _experimentationService;
+
+        [ImportingConstructor]
+        public AnalyzerVsixSuggestedActionCallback(VisualStudioWorkspace workspace,
+            IForegroundNotificationService foregroundNotificationService,
+            [ImportMany] IEnumerable<Lazy<IAsynchronousOperationListener, FeatureMetadata>> asyncListeners,
+            SVsServiceProvider serviceProvider)
+        {
+            _workspace = workspace;
+            _foregroundNotificationService = foregroundNotificationService;
+            _operationListener = new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.ABTest);
+            _serviceProvider = serviceProvider;
+        }
+
+        public void OnSuggestedActionExecuted(SuggestedAction action)
+        {
+            AssertIsForeground();
+
+            // Initialize the experimentation service if it hasn't yet been fetched
+            if (_experimentationService == null)
+            {
+                _experimentationService = _workspace.Services.GetRequiredService<IExperimentationService>();
+            }
+
+            // If we haven't yet checked to see if the VSIX is installed, check now
+            if (_installStatus == FxCopInstallStatus.Unchecked)
+            {
+                var vsShell = _serviceProvider.GetService(typeof(SVsShell)) as IVsShell;
+                ErrorHandler.ThrowOnFailure(vsShell.IsPackageInstalled(FxCopAnalyzersPackageGuid, out int installed));
+                _installStatus = installed != 0 ? FxCopInstallStatus.Installed : FxCopInstallStatus.NotInstalled;
+            }
+
+            // Only proceed if the VSIX isn't installed
+            if (_installStatus == FxCopInstallStatus.Installed)
+            {
+                return;
+            }
+
+            lock (_scheduleLock)
+            {
+                if (!_checksRunning && !_infoBarShown)
+                {
+                    _checksRunning = true;
+                    Task.Run((Action)SuggestedActionOccurred);
+                }
+            }
+        }
+
+        private void SuggestedActionOccurred()
+        {
+            // We don't need to do this work on the UI thread
```

Comment:
I assumed that we wouldn't want to do this on the UI thread. If we can just do this all on the UI thread, I'll move to that, as that'll make it all simpler.